### ISSUE TITLE
[release 1.6] Infinite loop in subtype checking code

### ISSUE THREAD

**Issue by olynch** [JuliaLang/julia#46929]
I get an infinite loop that prints out:

```
exists_subtype at /buildworker/worker/package_linux64/build/src/subtype.c:1426 [inlined]
forall_exists_subtype at /buildworker/worker/package_linux64/build/src/subtype.c:1454
subtype_in_env_ at /buildworker/worker/package_linux64/build/src/subtype.c:1909
subtype_bounds_in_env at /buildworker/worker/package_linux64/build/src/subtype.c:1919 [inlined]
try_subtype_in_env at /buildworker/worker/package_linux64/build/src/subtype.c:2194 [inlined]
intersect_var at /buildworker/worker/package_linux64/build/src/subtype.c:2307
intersect at /buildworker/worker/package_linux64/build/src/subtype.c:3015
intersect_union at /buildworker/worker/package_linux64/build/src/subtype.c:2164
intersect at /buildworker/worker/package_linux64/build/src/subtype.c:3028
intersect_all at /buildworker/worker/package_linux64/build/src/subtype.c:3190
intersect_aside at /buildworker/worker/package_linux64/build/src/subtype.c:2141
var_lt at /buildworker/worker/package_linux64/build/src/subtype.c:589
subtype at /buildworker/worker/package_linux64/build/src/subtype.c:1256
exists_subtype at /buildworker/worker/package_linux64/build/src/subtype.c:1426 [inlined]
forall_exists_subtype at /buildworker/worker/package_linux64/build/src/subtype.c:1454
subtype_in_env_ at /buildworker/worker/package_linux64/build/src/subtype.c:1909
subtype_bounds_in_env at /buildworker/worker/package_linux64/build/src/subtype.c:1919 [inlined]
try_subtype_in_env at /buildworker/worker/package_linux64/build/src/subtype.c:2194 [inlined]
....
```

when I run the following command:

```
julia --project=. MFE.jl
```

in [this branch](https://github.com/olynch/Catlab.jl/tree/mfe-infinite-subtype-loop). I managed to work around this error with [this commit](https://github.com/AlgebraicJulia/Catlab.jl/pull/665/commits/adfe671981c99892969a0e849fd02a7807716140). Basically, my understanding of this bug is that it happens when the type of the argument to `copy` is *partially* known; if it is fully known or not known at all, then dispatch works correctly, but somehow the partial knowledge that comes from `copy` being called with a type that's partially inferred causes the julia compiler to freak out.

**Comment by Keno** [comment]
I can't reproduce this on master. Could you try the nightly and see if the issue persists?

**Comment by olynch** [comment]
I'm so sorry; I should have specified that this only happens on 1.6: it's fixed in 1.7 and 1.8.

**Comment by vtjnash** [comment]
There has been a lot of changes to fix subtyping problems since then, and I am not sure we will end up backporting them fully
